dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/stack/transport/netty/SipStreamHandler.java

```diff
@@ -0,0 +1,112 @@
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+import jain.protocol.ip.sip.ListeningPoint;
+
+public class SipStreamHandler extends SimpleChannelInboundHandler<SipMessageByteBuffer> {
+
+    private static final TraceComponent tc = Tr.register(SipStreamHandler.class);
+
+    final private AttributeKey<SipTcpInboundConnLink> attrKey = AttributeKey.valueOf("SipTcpInboundConnLink");
+
+    /**
+     * Called when a new connection is established
+     */
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(this, tc, "channelActive", ctx.channel().remoteAddress() + " connected");
+        }
+
+        ListeningPoint lp = SipHandlerUtils.getListeningPoint(ctx.channel().localAddress(), ListeningPoint.TRANSPORT_TCP);
+        SipInboundChannel inboundChannel = null;
+
+        if (lp != null) {
+            inboundChannel = SIPConnectionFactoryImplWs.instance().getInboundChannels().get(lp);
+        }
+
+        if (inboundChannel != null) {
+            Attribute<SipTcpInboundConnLink> attr = ctx.channel().attr(attrKey);
+            attr.setIfAbsent(new SipTcpInboundConnLink(inboundChannel, ctx.channel()));
+        } else {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "channelActive",
+                        "could not associate an incoming connection with a SIP channel");
+            }
+        }
+    }
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, SipMessageByteBuffer msg) throws Exception {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(this, tc, "channelRead0",
+                    ctx.channel() + ". [" + msg.getMarkedBytesNumber() + "] bytes received");
+        }
+        Attribute<SipTcpInboundConnLink> attr = ctx.channel().attr(attrKey);
+        SipTcpInboundConnLink connLink = attr.get();
+
+        if (connLink != null) {
+            connLink.complete(msg);
+        } else {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "channelRead0", "could not associate an incoming messafe with a SIP channel");
+            }
+        }
+
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(this, tc, "channelInactive", ctx.channel().remoteAddress() + " has been disonnected");
+        }
+
+        Attribute<SipTcpInboundConnLink> attr = ctx.channel().attr(attrKey);
+        SipTcpInboundConnLink connLink = attr.get();
+        // clean up from connections table
+        if (connLink != null) {
+            connLink.destroy(null);
```

Comment:
Can we have a destroy(); as opposed to destroy(null)?

Reply:
done